<a href="https://colab.research.google.com/github/pypankj30/Rebounce-Applied-AI-and-Analytics-Codes/blob/main/Python_301.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Assignment MP3- Clean Analyse Summarise**

**Part 1 - Load and understand the data**

Importing pandas and numpy

In [1]:
import pandas as pd
import numpy as np

Loading data directly from web so that nothing loses when runtime restarts

In [2]:
url = "https://huggingface.co/datasets/aarav912/online-retail/resolve/main/online_retail.csv"
df = pd.read_csv(url)

Displaying first few rows


In [3]:
df.duplicated().sum()

np.int64(5268)

Checking the shape - how many rows and columns?

In [4]:
print(df.shape)
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

(541909, 8)
Number of rows: 541909
Number of columns: 8


Listing column names

In [5]:
print(df.columns)

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')


Checking data types of each column

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Counting missing values in every column

In [7]:
df.isna().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


Counting the duplicated rows

In [8]:
df.duplicated().sum()

np.int64(5268)

Description and CustomerID have null values , so they need cleaning

**Part 2- Clean the data**

In [9]:
#copy of clean original data
clean_df = df.copy()
print(clean_df.shape)


(541909, 8)


Keeping missing customerID becuase dropping them could impact overall data analysis as they are large in number

Checking data type of InvoiceDate

In [10]:
clean_df['InvoiceDate'].dtype

dtype('O')

Converting it to proper date time column from text

In [11]:
pd.to_datetime(clean_df['InvoiceDate'])

,InvoiceDate
0,2010-12-01 08:26:00
1,2010-12-01 08:26:00
2,2010-12-01 08:26:00
3,2010-12-01 08:26:00
4,2010-12-01 08:26:00
...,...
541904,2011-12-09 12:50:00
541905,2011-12-09 12:50:00
541906,2011-12-09 12:50:00
541907,2011-12-09 12:50:00


Dropping exact duplicates

In [12]:
clean_df = clean_df.drop_duplicates()
print(clean_df.shape)

(536641, 8)


Removing invalid transactions

In [13]:
clean_df = clean_df[(clean_df['Quantity'] > 0) & (clean_df['UnitPrice'] > 0)]
print(clean_df.shape)

(524878, 8)


Creating a new sales column

In [14]:
clean_df['Sales'] = clean_df['Quantity'] * clean_df['UnitPrice']
print(clean_df.shape)

(524878, 9)


**Part 3- NumPy analysis**

In [15]:
print(f"mean of sales: {np.mean(clean_df['Sales']):.2f}")
print(f"median of sales: {np.median(clean_df['Sales']):.2f}")
print(f"minimum of sales:{np.min(clean_df['Sales']):.2f}")
print(f"maximum of sales: {np.max(clean_df['Sales']):.2f}")
print(f"standard deviation of sales:{np.std(clean_df['Sales']):.2f}")

mean of sales: 20.28
median of sales: 9.92
minimum of sales:0.00
maximum of sales: 168469.60
standard deviation of sales:271.69


mean is almost twice of median , which tells us that sales are right skewed due to high value outliers

**Part 4- Pandas filtering**

1.transactions having sales above 500

In [16]:
print(f"Number of transactions having sales above 500: {clean_df[clean_df['Sales'] > 500].shape[0]}")

Number of transactions having sales above 500: 1155


10 largest transactions by sales?

In [17]:
largest_transactions = clean_df.sort_values('Sales', ascending=False)
print(f"largest transactions are:")
largest_transactions.head(10)

largest transactions are:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.60
61619,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,United Kingdom,77183.60
222680,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.50,15098.0,United Kingdom,38970.00
15017,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,United Kingdom,13541.33
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom,11062.06
173382,551697,POST,POSTAGE,1,2011-05-03 13:46:00,8142.75,16029.0,United Kingdom,8142.75
348325,567423,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,1412,2011-09-20 11:05:00,5.06,17450.0,United Kingdom,7144.72
160546,550461,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-04-18 13:20:00,2.10,15749.0,United Kingdom,6539.40
52711,540815,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-01-11 12:55:00,2.10,15749.0,United Kingdom,6539.40
421601,573003,23084,RABBIT NIGHT LIGHT,2400,2011-10-27 12:11:00,2.08,14646.0,Netherlands,4992.00


Transactions from Australia

In [18]:
print(f"Number of transactions from Australia: {clean_df[clean_df['Country'] == 'Australia'].shape[0]}")

Number of transactions from Australia: 1181


Transactions having quantity above 1000?

In [19]:
transaction_above_1000 = clean_df[clean_df['Quantity'] > 1000]
print(f"Number of transactions having quantity above 1000: {transaction_above_1000.shape[0]}")
transaction_above_1000.head(10)

Number of transactions having quantity above 1000: 106


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
4850,536809,84950,ASSORTED COLOUR T-LIGHT HOLDER,1824,2010-12-02 16:48:00,0.55,15299.0,United Kingdom,1003.20
4945,536830,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,2880,2010-12-02 17:38:00,0.18,16754.0,United Kingdom,518.40
4946,536830,21915,RED HARMONICA IN BOX,1400,2010-12-02 17:38:00,1.06,16754.0,United Kingdom,1484.00
6365,536890,17084R,ASSORTED INCENSE PACK,1440,2010-12-03 11:48:00,0.16,14156.0,EIRE,230.40
16435,537659,22189,CREAM HEART CARD HOLDER,1008,2010-12-07 16:43:00,2.31,18102.0,United Kingdom,2328.48
16436,537659,22188,BLACK HEART CARD HOLDER,1008,2010-12-07 16:43:00,2.31,18102.0,United Kingdom,2328.48
19871,537899,22328,ROUND SNACK BOXES SET OF 4 FRUITS,1488,2010-12-09 10:44:00,2.55,12755.0,Japan,3794.40
19980,537981,22492,MINI PAINT SET VINTAGE,1394,2010-12-09 11:35:00,0.55,17857.0,United Kingdom,766.70
25920,538420,17096,ASSORTED LAQUERED INCENSE HOLDERS,1728,2010-12-12 12:03:00,0.17,12875.0,United Kingdom,293.76
32671,539101,22693,GROW A FLYTRAP OR SUNFLOWER IN TIN,2400,2010-12-16 10:35:00,0.94,16029.0,United Kingdom,2256.00


Yes it seems believable because retailer many customers as wholesalers

**Part 5- GroupBy and aggregation**

In [20]:
summary_table = clean_df.groupby('Country').agg(
    Num_transactions=('InvoiceNo', 'count'),
    Total_sales=('Sales', 'sum'),
    Average_sales=('Sales', 'mean')
).reset_index()
print(summary_table)

                 Country  Num_transactions  Total_sales  Average_sales
0              Australia              1181   138453.810     117.234386
1                Austria               398    10198.680      25.624824
2                Bahrain                18      754.140      41.896667
3                Belgium              2031    41196.340      20.283772
4                 Brazil                32     1143.600      35.737500
5                 Canada               151     3666.380      24.280662
6        Channel Islands               747    20440.540      27.363507
7                 Cyprus               603    13502.850      22.392786
8         Czech Republic                25      826.740      33.069600
9                Denmark               380    18955.340      49.882474
10                  EIRE              7879   283140.520      35.936098
11    European Community                60     1300.250      21.670833
12               Finland               685    22546.080      32.913985
13    

which 5 countries have highest total sales

In [21]:
print(f"5 countries with the highest total sales:")
summary_table.sort_values('Total_sales', ascending=False).head(5)

5 countries with the highest total sales:


,Country,Num_transactions,Total_sales,Average_sales
36,United Kingdom,479985,9001744.094,18.754220
24,Netherlands,2359,285446.340,121.003111
10,EIRE,7879,283140.520,35.936098
14,Germany,9025,228678.400,25.338327
13,France,8392,209625.370,24.979191


Country with highest average transaction value

In [22]:
print(f"Country with the highest average transaction value:")
summary_table.sort_values('Average_sales', ascending=False).head(1)

Country with the highest average transaction value:


,Country,Num_transactions,Total_sales,Average_sales
24,Netherlands,2359,285446.34,121.003111


Number of transactions of top selling country

In [23]:
print(f"Number of transactions of top selling country:")
summary_table.sort_values('Num_transactions', ascending=False).head(1)

Number of transactions of top selling country:


,Country,Num_transactions,Total_sales,Average_sales
36,United Kingdom,479985,9001744.094,18.75422


**Part 6- Merge and region analysis**

Turning dictionary into small dataframe

In [24]:
region_lookup = {
"United Kingdom": "Europe",
"France": "Europe",
"Germany": "Europe",
"Spain": "Europe",
"Netherlands": "Europe",
"Belgium": "Europe",
"Switzerland": "Europe",
"Portugal": "Europe",
"Australia": "Oceania",
"Japan": "Asia",
"USA": "Americas",
}
region_lookup_df=pd.DataFrame(list(region_lookup.items()), columns=['Country', 'Region'])
region_lookup_df.head()

,Country,Region
0,United Kingdom,Europe
1,France,Europe
2,Germany,Europe
3,Spain,Europe
4,Netherlands,Europe


Merging into country summary table

In [25]:
country_summary_table = pd.merge(summary_table, region_lookup_df, on='Country', how='left')
country_summary_table['Region'] = country_summary_table['Region'].fillna('Other')
country_summary_table.head()

,Country,Num_transactions,Total_sales,Average_sales,Region
0,Australia,1181,138453.81,117.234386,Oceania
1,Austria,398,10198.68,25.624824,Other
2,Bahrain,18,754.14,41.896667,Other
3,Belgium,2031,41196.34,20.283772,Europe
4,Brazil,32,1143.60,35.737500,Other


Grouping by region and finding total sales

In [26]:
summary_region = country_summary_table.groupby('Region')['Total_sales'].sum().reset_index()
print(f"Total sales by region:")
summary_region



Total sales by region:


,Region,Total_sales
0,Americas,3580.390
1,Asia,37416.370
2,Europe,9918999.754
3,Oceania,138453.810
4,Other,543660.480


Region with highest total sales

In [27]:
print("Region with the highest total sales:")
summary_region.sort_values('Total_sales', ascending=False).head(1)

Region with the highest total sales:


,Region,Total_sales
2,Europe,9918999.754


**Part 7 - Answer the questions**

1. Which country generated the highest total sales?

   United kingdom with 9001744.094	total sales

2. Which country had the highest average transaction value?

     Netherlands with 121.003111 average transaction value


3. Which region performed best?

     Europe with 9918999.754 total sales

4. What is one interesting pattern you noticed in data ?

     United kingdom with highest 9001744.094 total sales but lowest average value 18.75422 and highest transactions 479985 and Netherlands with highest 121.003111 average transaction value and lowest transactions 2359 and 285446.34 total sales .

5. What is one data quality problem you ran into, and how did you handle it ?

   There were 135080 missing customerIds which account for roughly 25% ,removing them could impact overall analysis , so kept them.
   And there were 5268 identical duplicate rows , removed them to have better data analysis.

**Part 8- Final insight**

This online retail data set shows while United Kingdom had the highest transactions but low average and Netherlands on the other hand having lowest transactions have highest average transaction value. The mean is almost twice the median which shows the right skewedness of data due to outliers which does not represent true picture. Duplication in data could impact overall analysis , so it had to be cleaned.
